bootstrap/deploy_job.yaml

```diff
@@ -0,0 +1,31 @@
+# Deploy kubeflow to namespace "kubeflow"
+
+# TODO(inc0): Create rbac role bindings
+---
+apiVersion: v1
```

Comment:
Running in cluster as a job seems like it creates RBAC issues which is one of the current problems I'd like the bootstrapper to deal with.
https://github.com/kubeflow/kubeflow/blob/master/user_guide.md#rbac-clusters

Deploying Kubeflow requires high level permissions e.g. because you need to create things like RBAC roles. You also might want to create keys to use as secrets for GCP/S3.

So if you run on cluster asynchronously then you're forcing users to go through the hurdle of creating appropriate service accounts and provisioning them; which adds the kind of complexity we're trying to avoid.

The bootstraper aims to avoid this by being designed to run interactively.  If it runs interactively then the bootstraper can use the user's credentials. This allows us to use the end user's credentials to boot strap the cluster.

For example, in the current execution mode the bootstrapper runs locally in docker on the user's computer so we can have the user go through whatever login flow is neccessary to get a credential and then use that. So we can use the user's credential to create cluster roles, Cloud secrets etc.

We can eventually allow this to be run on cluster e.g. via kubectl exec -ti but we need to be careful about how we handle the user credential because we wouldn't want to store the user credential in a pod unless the pod is properly locked down.

One solution is to make it a web app. In that case the refresh token can be stored client side and the web app can just get short lived access tokens to forward on to various services.

Reply:
Current pattern is to create RBAC bindings in bootstrap. Look at this: https://github.com/kubeless/kubeless/releases/download/v0.6.0/kubeless-v0.6.0.yaml 

I think kubeless is good example of what I'd like to see - they specify required ClusterRoles and make it work.

Reply:
@inc0 @jlewi We've been doing a lot of prototyping on separating bootstrapper into specific phases:

- create cluster level resources: clusterroles, clusterbindings, namespace, pod security contexts, service accounts, PVs. In short, isolate ksonnet libs into those which require cluster-admin equivalent RBAC roles. 
- create a single-signon that uses IAP or other authentication providers. This auth provider will use web-token authentication (we've experimented with oauth/openid tokens as well) and will likely be based on or use https://github.com/appscode/guard. 
- define RBAC roles based on the provider. For example roles for 'data scientist' users, leads, etc. One thing guard does is import github users in an organization including teams. Depending on the auth provider, a RBAC mapping of teams could be automated so that data scientists have a 'zero-admin' experience. 

We see bootstrapper as having to address 2 concerns:  'cloud-providers' and 'authentication providers'.  Each are distinct and could be bundled as ksonnet libs similarly to IAP. Having a framework in bootstrapper that can mix and match cloud providers with auth providers is how we've been thinking about this problem. Part of the difficulty we've been having is under GKE it's not straightforward to update an api-server with the equivalent of boot flags like --authentication-token-webhook-config-file or --oidc-issuer-url. We've done a fair amount of prototyping using kops, dex, and guard. @jlewi it looks like currently there is an option to use cloud endpoints and ESP to delegate to a provider that does IAP under GKE. We've gone down this path as well to use other providers. We're working on a document you both can review (and others) but wanted to have enough concrete prototyping done to have answers for questions. Is there a particular preference for how the bootstrapper is implemented? It seems like the go implementation has the most potential to get at features not normally exposed in CLI tools like ks, gcloud, kubectl, etc.

Reply:
Also another problem is on prem which can be any mixture of this. I'm not sure if we should go as far as try to autodetect this. I's going to be really tricky as every on prem is different.

RBAC is slightly different problem tho (imho). One is managing users (I'm user X) and another is managing roles (user X is allowed to do Y). 1st is external to k8s (but also we shouldn't solve this ourselves, there are projects doing this), another is just k8s - RBAC.

In any case I don't think we should discuss it here, it's broader discussion that we need to have and will affect much more (auth to jupyterhub?). I'd just say we that we can for now create namespace and rolebindings to this namespace, which user gets there roles - let's discuss that separately.

Reply:
@inc0 my apologies - i'm guilty of hijacking the context of your proposal! Allow me to reply to Jeremy and we can take this up separately. I do agree with your point that we shouldn't embed a user db or take on user management. 

@jlewi yes the idea is to bundle by cloud provider and auth provider which should reduce a number of parameters. And we could probably make sure the deployer has appropriate privileges.
> Why is creating RBAC roles a separate phase? RBAC roles for datascientists/users/ etc... should be defined in your ksonnet app and deployed when you deploy Kubeflow.

the ClusterRoles can be defined on deployment but something like github which (as an auth provider) can return a mapping of teams and individuals could be done in a later phase after the auth provider was defined. This is similar to IAP where it's run after core packages.
> But there are likely some things that you can't manage as K8s resources. For example, on GCP you might want to create a static IP. So in bootstrapper we'd just make an RPC to the appropriate GCP API.

yes calls to the GCP API (ie service and endpoint apis) can be isolated by cloud provider. There are no doubt similar calls to AWS IAM. 

You both may be interested in reading this proposal [AuthN/AuthZ in kubeapps](https://github.com/prydonius/kubeapps/blob/2df463b4a74919bb13e4893394a8741969c0c120/docs/design-proposals/authentication-and-authorization.md) which has some similarities in approach and I thought was well written.

Reply:
> Why is creating RBAC roles a separate phase? RBAC roles for datascientists/users/ etc... should be defined in your ksonnet app and deployed when you deploy Kubeflow.

I agree with @jlewi that a data scientist or ML end user should not have to understand RBAC, ClusterRoles, or policies to get up and running with a "production"-ready research cluster.

Reply:
@inc0 @nickchase @willingc 

What if the experience was?

kubectl create -f  kubeflow_cli.yaml
kubectl exec -ti kubeflow-cli-0

Then in the resulting shell run
/opt/kubeflow/quick_start.sh
(I would be ok if quick_start.sh  ran automatically on startup and then the pod entered a loop forever to allow kubectl exec)

What do you people think?

There's a couple things I like about this approach

1. We can use a PD to store the ksonnet app
1. User can start a shell to interact with the tool chain (kubectl/ks etc...) but it doesn't get in the way.

I think this allows us to introduce concepts/tools gradually. Users don't need to know ksonnet just to deploy Kubeflow.

But if they want to customize various parameters we can tell them how to do this using `ks` without having to deal with installation.

The ksonnet app is also available to them so they can push it to source control or do more advanced manipulations by editing the files directly.

Reply:
To clarify a bit on @jlewi's suggestion to make sure that I am understanding it correctly:

1. `kubectl create -f kubeflow_cli.yaml` where `kubeflow_cli.yaml` has defaults so a data scientist does not have to decide them to deploy.
2. `kubectl exec -ti kubeflow-cli-0` opens a shell for the user to interact with `kubectl` or `ks`.

a) If the user doesn't wish to interact with the shell, what is displayed to them?
b) Kubeflow is fully deployed at this point (i.e. I can open a notebook or start a ML job)
c) `ks` is still available if I am familiar with how use it to make modifications

Forgive me, I'm not sure what "PD" stands for. Assuming its some sort of persistent storage or persistent disk (PD?).

Reply:
> kubectl create -f kubeflow_cli.yaml where kubeflow_cli.yaml has defaults so a data scientist does not have to decide them to deploy.

Correct in that user doesn't have to specify anything.

> a) If the user doesn't wish to interact with the shell, what is displayed to them?
I think it depends on what we decide to do.
We could either require the user to always login via kubectl exec and then do "quick_start.sh" to deploy Kubeflow with automatic defaults.

Or we could just have kubectl crete -f kubeflow_cli.yaml automatically run quick_start.sh so user would only have to start a shell and use kubectl/ks if they weren't happy with the auto-configured deployment.

> Forgive me, I'm not sure what "PD" stands for. Assuming its some sort of persistent storage or persistent disk (PD?). 

Correct; persistent disk.

I'm basically just suggesting a couple modifications to @inc0's original PR

1. We should by default store the ksonnet app on the K8s cluster using the cluster default PVC (if cluster doesn't have a default, that could be handled as an exceptional case requiring the user to make a change to the YAML file).

1. Instead of running the container as a job and exiting, we should have the container enter an idle loop so that users can just do `kubectl exec` to start a shell and interact with the app and kubeflow deployment after the initial deployment is complete

1. I don't think we should expose ksonnet parameters in kubeflow_cli.yaml. Instead, we should try to make the bootstrap go program more intelligent and pick better defaults for the ksonnet app. If users want to customize this deployment then we should tell customers how to do this using ksonnet.

Reply:
@jlewi for 2. above could `kubectl exec ...` also act as a kind of [REPL](https://en.wikipedia.org/wiki/Read%E2%80%93eval%E2%80%93print_loop)? That is the data scientist doesn't have to explicitly enter a shell but could do something like `kubectl exec ... tfjob start job1`. This would operate as a type of CLI if the bootstrap go program is given the REPL  `tfjob start job1` and would know that it needs to call the ks api to generate and apply the tfjob package. Maybe this is what you had in mind with 3. with making the go program more intelligent and could accommodate customizations.

Reply:
@inc0 Did you see my [comment above](https://github.com/kubeflow/kubeflow/pull/673#discussion_r182546175)

@kkasravi That's an interesting idea. It seems like that would just work since `kubectl exec` allows executing  binaries in a container that is already running.  So it would work for any CLI we install.